web/autolinker.js

```diff
@@ -0,0 +1,156 @@
+/* Copyright 2025 Mozilla Foundation
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {
+  AnnotationBorderStyleType,
+  AnnotationType,
+  createValidAbsoluteUrl,
+  Util,
+} from "pdfjs-lib";
+import { getOriginalIndex, normalize } from "./pdf_find_controller.js";
+
+function DOMRectToPDF({ width, height, left, top }, pdfPageView) {
+  if (width === 0 || height === 0) {
+    return null;
+  }
+
+  const pageBox = pdfPageView.textLayer.div.getBoundingClientRect();
+  const bottomLeft = pdfPageView.getPagePoint(
+    left - pageBox.left,
+    top - pageBox.top
+  );
+  const topRight = pdfPageView.getPagePoint(
+    left - pageBox.left + width,
+    top - pageBox.top + height
+  );
+
+  return Util.normalizeRect([
+    bottomLeft[0],
+    bottomLeft[1],
+    topRight[0],
+    topRight[1],
+  ]);
+}
+
+function calculateLinkPosition(range, pdfPageView) {
+  const rangeRects = range.getClientRects();
+  if (rangeRects.length === 1) {
+    return { rect: DOMRectToPDF(rangeRects[0], pdfPageView) };
+  }
+
+  const rect = [Infinity, Infinity, -Infinity, -Infinity];
+  const quadPoints = [];
+  let i = 0;
+  for (const domRect of rangeRects) {
+    const normalized = DOMRectToPDF(domRect, pdfPageView);
+    if (normalized === null) {
+      continue;
+    }
+
+    quadPoints[i] = quadPoints[i + 4] = normalized[0];
+    quadPoints[i + 1] = quadPoints[i + 3] = normalized[3];
+    quadPoints[i + 2] = quadPoints[i + 6] = normalized[2];
+    quadPoints[i + 5] = quadPoints[i + 7] = normalized[1];
+
+    rect[0] = Math.min(rect[0], normalized[0]);
+    rect[1] = Math.min(rect[1], normalized[1]);
+    rect[2] = Math.max(rect[2], normalized[2]);
+    rect[3] = Math.max(rect[3], normalized[3]);
+
+    i += 8;
+  }
+  return { quadPoints, rect };
+}
+
+function createLinkAnnotation({ url, index, length }, pdfPageView, id) {
+  const convertedMatch = pdfPageView._textHighlighter._convertMatches(
+    [index],
+    [length]
+  )[0];
+
+  const range = new Range();
+  range.setStart(
+    pdfPageView._textHighlighter.textDivs[convertedMatch.begin.divIdx]
+      .firstChild,
+    convertedMatch.begin.offset
+  );
+  range.setEnd(
+    pdfPageView._textHighlighter.textDivs[convertedMatch.end.divIdx].firstChild,
+    convertedMatch.end.offset
+  );
```

Comment:
Can't this be shortened/simplified a little bit?
```suggestion
  const { _convertMatches, textDivs } = pdfPageView._textHighlighter;
  const [convertedMatch] = _convertMatches([index], [length]);

  const range = new Range();
  range.setStart(
    textDivs[convertedMatch.begin.divIdx].firstChild,
    convertedMatch.begin.offset
  );
  range.setEnd(
    textDivs[convertedMatch.end.divIdx].firstChild,
    convertedMatch.end.offset
  );
```

Reply:
Doing this didn't work because `_convertMatches` needs `this` and binding wouldn't have made the line much shorter. I've improved this part of the code in another way, let me know how it looks.